Dockerfile

```diff
@@ -1,33 +1,72 @@
+# https://docs.docker.com/go/dockerfile-reference/
+
+# Want to help us make this template better? Share your feedback here: https://forms.gle/ybq9Krt8jtBL3iCk7
+
+ARG NODE_VERSION=20
+
+################################################################################
+# Use node image for base image for all stages.
+FROM node:${NODE_VERSION}-alpine as base
+
+# Set working directory for all build stages.
 WORKDIR /app
 
-COPY package*.json ./
 
-RUN npm ci
+################################################################################
+# Create a stage for installing production dependecies.
+FROM base as deps
 
-COPY . .
+# Download dependencies as a separate step to take advantage of Docker's caching.
+# Leverage a cache mount to /root/.npm to speed up subsequent builds.
+# Leverage bind mounts to package.json and package-lock.json to avoid having to copy them
+# into this layer.
+RUN --mount=type=bind,source=package.json,target=package.json \
+  --mount=type=bind,source=package-lock.json,target=package-lock.json \
+  --mount=type=cache,target=/root/.npm \
+  npm ci --omit=dev
 
-# Build the application
+################################################################################
+# Create a stage for building the application.
+FROM deps as build
+
+# Download additional development dependencies before building, as some projects require
+# "devDependencies" to be installed to build. If you don't need this, remove this step.
+RUN --mount=type=bind,source=package.json,target=package.json \
+  --mount=type=bind,source=package-lock.json,target=package-lock.json \
+  --mount=type=cache,target=/root/.npm \
+  npm ci
+
+# Copy the rest of the source files into the image.
+COPY . .
+# Run the build script.
 RUN npm run build
 
-# Production stage
-FROM node:20
+################################################################################
+# Create a new stage to run the application with minimal runtime dependencies
+# where the necessary files are copied from the build stage.
+FROM base as final
 
-WORKDIR /app
+# Use production node environment by default.
+ENV NODE_ENV production
+
+# Run the application as a non-root user.
+USER node
 
-COPY --from=builder /app/dist ./dist
-COPY package*.json ./
+# Copy package.json so that package manager commands can be used.
+COPY package.json .
 
-RUN npm ci --omit=dev
+# Copy the production dependencies from the deps stage and also
+# the built application from the build stage into the image.
+COPY --from=deps /app/node_modules ./node_modules
+COPY --from=build /app/dist ./dist
 
-# We want jq and curl in the final image, but we don't need the support files
-RUN apt-get update && \
-	apt-get install -y jq curl tini && \
-	apt-get clean && \
-	rm -rf /usr/share/doc /usr/share/man /usr/share/zsh
 
+# Expose the port that the application listens on.
 EXPOSE 3000
 
-ENTRYPOINT ["/usr/bin/tini", "--", "npm", "run", "start:prod"]
+# Run the application.
+CMD node dist/apps/api/main.js
```

Comment:
1. Do we not want to use `tini` any more? I find that apps shut down better with it...
2. Wouldn't mind talking through the logic of `ENTRYPOINT` vs `CMD`; I know it's just taking the auto-generated approach, but there is a not insignificant difference, and I'd like to be intentional about such changes as opposed to just "take the auto-generated stuff"

Reply:
I think in this case as we have no command line arguments it should be the same right?

Reply:
1. I'd like to be consistent in how we build the Docker images across all the Gateway services
2. I prefer to allow for the option, in case we want to pass CLI options in the future...

Also, what about the `tini` question? Seems like it's much more well-behaved in processing signals and passing them to the app, makes for a much nicer shutdown experience, at least when doing development.

Reply:
I can swap it back to `tini`, although I'd like to run with just node instead of npm still

Reply:
Swapped back to tini on both Dockerfile and dev.Dockerfile